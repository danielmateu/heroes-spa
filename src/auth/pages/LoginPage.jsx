import { useContext } from "react"
import { useNavigate } from "react-router-dom"
import { AuthContext } from "../context/AuthContext"


export const LoginPage = () => {

  

  const {login} = useContext(AuthContext)
  const navigate = useNavigate()

  const onLogin = () =>{

    const lastPath = localStorage.getItem('lastPath') || '/';

    login('Daniel Mateu');

    navigate(lastPath, { 
      replace: true,
    })
  }
  return (
    <>
    <div className="container mt-4">
    <h1>Login</h1>

    <hr />

    {/* <div className="input-group input-group-default mb-3">
    <span 
    className="input-group-text" 
    id="inputGroup-sizing-default"
    >Name
    </span>

    <input 
    className="form-control "
    type="text" 
    placeholder="Username"
    aria-label="Sizing example input" 
    aria-describedby="inputGroup-sizing-default"
    />
    </div> */}

    {/* <div className="input-group input-group-default mb-3">
    <span 
    className="input-group-text" 
    id="inputGroup-sizing-default"
    >Password
    </span>

    <input 
    className="form-control "
    type="password" 
    placeholder="Insert Password"
    aria-label="Sizing example input" 
    aria-describedby="inputGroup-sizing-default"
    />
    </div> */}
    
    <button 
    onClick={onLogin}
    className="btn btn-outline-primary">Login</button>

    </div>
    </>
  )
}
