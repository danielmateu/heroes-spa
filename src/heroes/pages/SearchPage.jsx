import queryString from 'query-string';
import { useLocation, useNavigate } from 'react-router-dom';
import { useForm } from '../../hooks/useForm'
import { HeroCard } from '../components/HeroCard';
import { getHeroesByName } from '../helpers';


export const SearchPage = () => {

  const navigate = useNavigate();
  const location = useLocation();

  const {q = ''} = queryString.parse(location.search);
  const heroes = getHeroesByName(q);

  const showSearch = (q.length === 0);
  const showError = (q.length > 0) && heroes.length === 0;

  const {searchText, onInputChange} = useForm({
    searchText:q
  })

  const onSearchSubmit = (e) =>{
    e.preventDefault();
    // if(searchText.trim().length <= 1) return;
    
    navigate(`?q=${searchText}`)
    
  }

  return (
    <>
    <h1 className='animate__animated animate__fadeInLeft mt-2'>Search</h1>   
  <hr />

  <div className="row">
    <div className="col-5">
      <h4>Searching</h4>
      <hr />
      <form onSubmit={onSearchSubmit} aria-label={'form'}>
        <input type="text"
        placeholder='search your hero'
        className='form-control'
        name='searchText'
        autoComplete='off'
        value={searchText}
        onChange={onInputChange} />
      </form>
      <button 
      className='btn btn-outline-info mt-2'
      onClick={onSearchSubmit}>Search</button>
    </div>
   
    <div className="col-7">
      <h4>Results</h4>
    <hr />
    {/* {
      (q === '') ? <div className="alert alert-primary">
      Search a Hero!
    </div> : (heroes.length === 0) && <div className="alert alert-danger">
      There are no results with <b>{q}</b>
    </div>
    } */}
    
    <div className="animate__animated animate__fadeIn alert alert-primary" aria-label='alert-primary' style={{display: showSearch ? '' :'none'}} >Search a Hero
    </div>

    <div className="animate__animated animate__fadeIn alert alert-danger" aria-label='alert-danger' style={{display:showError? '' : 'none'}} >There are no results with <b>{q}</b>
    </div>

    
      {
      heroes.map(hero => 
        <HeroCard key={hero.id} {...hero}/>
      )
    }
    
    

    {/* <HeroCard/> */}

    </div>
    
  </div>
    
    </>
  )
}
