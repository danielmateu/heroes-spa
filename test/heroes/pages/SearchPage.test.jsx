import { render, screen, fireEvent } from "@testing-library/react"
import { MemoryRouter } from "react-router-dom"
import { SearchPage } from "../../../src/heroes/pages/SearchPage"

const mockUseNavigate = jest.fn()

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useNavigate: () => mockUseNavigate 
}));


describe('Pruebas sobre Search Page', () => { 

    beforeEach(() => jest.clearAllMocks()); 

    test('Debe de mostrarse correctamente con valores por defecto', () => { 
        const {container} = render(
            <MemoryRouter>
                <SearchPage/>
            </MemoryRouter>
        );

        // screen.debug();

        expect(container).toMatchSnapshot();
    });

    test('Debe de mostrar a Batman y el input con el valor del queryString', () => {
        
        render(
            <MemoryRouter initialEntries={['/search?q=batman']}>
                <SearchPage/>
            </MemoryRouter>
        );

        // screen.debug();

        const input = screen.getByRole('textbox');
        expect(input.value).toBe('batman');

        const image = screen.getByRole('img');
        expect(image.src).toContain('/assets/heroes/dc-batman.jpg')

        const alert = screen.getByLabelText('alert-primary');
        // console.log(alert);

        expect(alert.style.display).toBe('none');

    });

    test('Debe de mostrar un error si no se encuentra el Heroe', () => {  

        render(
            <MemoryRouter initialEntries={['/search?q=me%20cago%20en%20dios']}>
                <SearchPage/>
            </MemoryRouter>
        );

        const alert = screen.getByLabelText('alert-danger');
        expect(alert.style.display).toBe('');

    })

    test('Debe de llamar el navigate a la pantallanueva', () => {  
        

        const inputValue = 'superman'

        render(
            <MemoryRouter initialEntries={['/search']}>
                <SearchPage/>
            </MemoryRouter>
        );

        const input = screen.getByRole('textbox');
        fireEvent.change(input, {target: {name: 'searchText', value: inputValue}})

        console.log(input.value);

        const form = screen.getByRole('form');
        fireEvent.submit(form);

        expect(mockUseNavigate).toHaveBeenCalledWith(`?q=${inputValue}`);

    })


})