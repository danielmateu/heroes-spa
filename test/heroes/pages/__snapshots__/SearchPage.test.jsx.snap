// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas sobre Search Page Debe de mostrarse correctamente con valores por defecto 1`] = `
<div>
  <h1
    class="animate__animated animate__fadeInLeft mt-2"
  >
    Search
  </h1>
  <hr />
  <div
    class="row"
  >
    <div
      class="col-5"
    >
      <h4>
        Searching
      </h4>
      <hr />
      <form
        aria-label="form"
      >
        <input
          autocomplete="off"
          class="form-control"
          name="searchText"
          placeholder="search your hero"
          type="text"
          value=""
        />
      </form>
      <button
        class="btn btn-outline-info mt-2"
      >
        Search
      </button>
    </div>
    <div
      class="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        aria-label="alert-primary"
        class="animate__animated animate__fadeIn alert alert-primary"
      >
        Search a Hero
      </div>
      <div
        aria-label="alert-danger"
        class="animate__animated animate__fadeIn alert alert-danger"
        style="display: none;"
      >
        There are no results with 
        <b />
      </div>
    </div>
  </div>
</div>
`;
